subprojects/core-api/src/main/java/org/gradle/api/execution/incremental/IncrementalInputs.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.execution.incremental;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.NonExtensible;
+import org.gradle.api.tasks.incremental.InputFileDetails;
+
+/**
+ * Provides access to any input files that need to be processed by an incremental work action.
+ *
+ * <p>
+ * An incremental work action is one that accepts a single {@link IncrementalInputs} parameter.
+ * The work action can then query what changed for an input property since the last execution to only process the changes.
+ *
+ * <pre class='autoTested'>
+ * class IncrementalReverseTask extends DefaultTask {
+ *     {@literal @}InputDirectory
+ *     def File inputDir
+ *
+ *     {@literal @}OutputDirectory
+ *     def File outputDir
+ *
+ *     {@literal @}TaskAction
+ *     void execute(IncrementalInputs inputs) {
+ *         if (!inputs.incremental)
+ *             project.delete(outputDir.listFiles())
+ *
+ *         inputs.getChanges(inputDir).each { change -&gt;
+ *             if (change.removed) {
+ *                 def targetFile = project.file("$outputDir/${change.file.name}")
+ *                 if (targetFile.exists()) {
+ *                     targetFile.delete()
+ *                 }
+ *             } else {
+ *                 def targetFile = project.file("$outputDir/${change.file.name}")
+ *                 targetFile.text = change.file.text.reverse()
+ *             }
+ *         }
+ *     }
+ * }
+ * </pre>
+ *
+ * <p>
+ * In the case where Gradle is unable to determine which input files need to be reprocessed, then all of the input files will be reported as added.
+ * Cases where this occurs include:
+ * <ul>
+ *     <li>There is no history available from a previous execution.</li>
+ *     <li>An non-file input property has changed since the previous execution.</li>
+ *     <li>One or more output files have changed since the previous execution.</li>
+ * </ul>
+ *
+ * @since 5.4
+ */
+@NonExtensible
```

Comment:
With the current implementation it is not, since we don't use the instantiator to create the object. So we don't need to signal to the code generation to not add the extension methods.

Reply:
I removed `@NonExtensible`.